### PR TITLE
astropy and numpy are now dependences, pyfits can be used.

### PR DESCRIPTION
**Pull request by montefra**
- [x] numpy is a dependancy, there is no need to check it
- [x] astropy is a dependancy, there is no need to check it
- [x] make it possible to use either astropy or pyfits fits representation. Will help for cases like #51

### CODE REVIEW

**Comment by montefra** [comment]
@ericmandel : I have began to add tests.

As far as you know, is there any way to run ds9 without the GUI?

**Comment by ericmandel** [comment]
> As far as you know, is there any way to run ds9 without the GUI?

@montefra You need an X server, but you can use the virtual Xvfb. See:

http://ds9.si.edu/doc/faq.html 

for details. @wjoye will have more recent experience with this than I.

**Comment by montefra** [comment]
Thank you for the pointer. I'll see if I can use this: it might be enough.
For now I just want to check that I can send and receive data.

**Comment by montefra** [comment]
there is a [pytest pluging for Xvfb](https://pypi.python.org/pypi/pytest-xvfb/).
I'll give a try in the next few days.

**Comment by ericmandel** [comment]
That would be pretty convenient!

**Comment by montefra** [comment]
@ericmandel : good news! ``pytest-xvfb`` works and I can test ds9. I have added tests for ``{set,get}_fits``. I will add similar tests for ``{set,get}_pyfits`` and then I will call this issue closed

File: pyds9/conftest.py [hunk 1]
```diff
@@ -36,3 +39,16 @@
...
+
+
+@pytest.fixture
+def test_data_dir():
+    '''Returns the name of the test data directory as a py.path.local object'''
+    this_file = py.path.local(__file__)
+    return this_file.dirpath('tests', 'data')
+
+
+@pytest.fixture
```
Comment (montefra) [inline pyds9/conftest.py]: This fixture returns the name of the test fits file

File: pyds9/pyds9.py [hunk 2]
```diff
@@ -27,6 +27,8 @@
 
 from astropy.extern import six
 from astropy.extern.six import BytesIO
+from astropy.io import fits
```
Comment (montefra) [inline pyds9/pyds9.py]: Astropy and numpy are dependencies, so there is no point in checking for them

File: pyds9/pyds9.py [hunk 3]
```diff
@@ -116,75 +118,92 @@ def get_xpans_ds9():
...
-except:
-    try:
-        import pyfits
-        if pyfits.__version__ >= '2.2':
-            ds9Globals["pyfits"] = 2
-        else:
-            ds9Globals["pyfits"] = 0
-    except:
-        ds9Globals["pyfits"] = 0
+    import pyfits
```
Comment (montefra) [inline pyds9/pyds9.py]: Check for the existence of ``pyfits``

File: pyds9/pyds9.py [hunk 4]
```diff
@@ -116,75 +118,92 @@ def get_xpans_ds9():
...
-            return numpy.int64
-        elif bitpix == -32:
-            return numpy.float32
-        elif bitpix == -64:
-            return numpy.float64
-        elif bitpix == -16:
-            return numpy.uint16
-        else:
-            raise ValueError('unsupported bitpix: %d' % bitpix)
+def _bp2np(bitpix):
```
Comment (montefra) [inline pyds9/pyds9.py]: Since numpy is always present, there is no need to guards ``_bp2np``and ``_np2bp``.

File: pyds9/pyds9.py [hunk 5]
```diff
@@ -610,184 +630,270 @@ def access(self):
...
-                    raise ValueError('requires pyfits.HDUList as input')
-                else:
-                    raise ValueError('requires astropy.HDUList as input')
-            # for python2 BytesIO and StringIO are the same
-            with contextlib.closing(BytesIO()) as newFitsFile:
-                hdul.writeto(newFitsFile)
-                newfits = newFitsFile.getvalue()
-                got = self.set('fits', newfits, len(newfits))
-                return got
+    def _ds9_fits_to_bytes(self):
```
Comment (montefra) [inline pyds9/pyds9.py]: Return the fits in ds9 as a byte stream

File: pyds9/pyds9.py [hunk 6]
```diff
@@ -610,184 +630,270 @@ def access(self):
...
+        Returns
+        -------
+        :class:`six.BytesIO`
+            file-like object containing the FITS data
+        '''
+        self._selftest()
+        imgData = self.get('fits')
+        return BytesIO(string_to_bytes(imgData))
+
+    def _hdulist_to_ds9_fits(self, hdul):
```
Comment (montefra) [inline pyds9/pyds9.py]: send a hdulist to ds9

File: pyds9/pyds9.py [hunk 7]
```diff
@@ -610,184 +630,270 @@ def access(self):
...
+        '''
+        self._selftest()
+        # for python2 BytesIO and StringIO are the same
+        with contextlib.closing(BytesIO()) as newFitsFile:
+            hdul.writeto(newFitsFile)
+            newfits = newFitsFile.getvalue()
+            success = self.set('fits', newfits, len(newfits))
+        return success
+
+    def get_fits(self):
```
Comment (montefra) [inline pyds9/pyds9.py]: get a fits from ds9 as an ``astropy`` HDUList

File: pyds9/pyds9.py [hunk 8]
```diff
@@ -610,184 +630,270 @@ def access(self):
...
-            This method is not defined because numpy in not installed.
-            """
-            raise ValueError('get_arr2np not defined (numpy not found)')
-
-        def set_np2arr(self):
-            """
-            This method is not defined because numpy in not installed.
-            """
-            raise ValueError('set_np2arr not defined (numpy not found)')
+    def set_fits(self, hdul):
```
Comment (montefra) [inline pyds9/pyds9.py]: send an ``astropy`` HDUList to ds9

File: pyds9/pyds9.py [hunk 9]
```diff
@@ -610,184 +630,270 @@ def access(self):
...
+        Raises
+        ------
+        ValueError
+            if the input is not an astropy HDUList
+        """
+        if not isinstance(hdul, fits.HDUList):
+            raise ValueError('The input must be an astropy HDUList')
+        return self._hdulist_to_ds9_fits(hdul)
+
+    def get_pyfits(self):
```
Comment (montefra) [inline pyds9/pyds9.py]: analogous to the ``{get,set}_fits`` methods, but for pyfits. If ``pyfits`` is not found, raise an attribute error. I have duplicated the methods to transparently allow setting a pyfits or an astropy fits.

File: pyds9/pyds9.py (hunk 7 above)
Comment (montefra) [reply pyds9/pyds9.py]: I just realised that I need to thing about backward compatibility. I'll think something

File: pyds9/pyds9.py (hunk 8 above)
Comment (montefra) [reply pyds9/pyds9.py]: I just realised that I need to thing about backward compatibility. I'll think something

File: pyds9/pyds9.py [hunk 10]
```diff
@@ -610,184 +630,270 @@ def access(self):
...
+        if not ds9Globals["pyfits"]:
+            raise AttributeError("'DS9' object has not attribute"
+                                 " 'set_pyfits'. The method is available only"
+                                 " if the package 'pyfits' is imported and"
+                                 " it's version >=2.2")
+        if not isinstance(hdul, pyfits.HDUList):
+            raise ValueError('The input must be a pyfits HDUList')
+        return self._hdulist_to_ds9_fits(hdul)
+
+    def get_arr2np(self):
```
Comment (montefra) [inline pyds9/pyds9.py]: The ``{get,set}_arr2np`` don't need to be guarded anymore

File: pyds9/tests/test_pyds9.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    assert len(ds9s) == 1
+    assert ds9_title in ds9s[0].target
+
+
+# def test_ds9(ds9_title):
+    # ''''''
+    # ds9 = pyds9.ds9_openlist(target=ds9_title)[0]
+
+
+def test_ds9_get_fits(ds9_title, test_fits):
```
Comment (montefra) [inline pyds9/tests/test_pyds9.py]: Load a fits file and then get it with the ``get_fits`` method. They should be the same

File: pyds9/tests/test_pyds9.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+
+@pytest.mark.xfail(raises=ValueError, reason='Not an astropy hdu')
+def test_ds9_set_fits_fail(ds9_title):
+    '''set_fits wants an astropy HDUList'''
+    ds9 = pyds9.ds9_openlist(target='*' + ds9_title + '*')[0]
+    ds9.set_fits('random_type')
+
+
+def test_ds9_set_fits(tmpdir, ds9_title, test_fits):
```
Comment (montefra) [inline pyds9/tests/test_pyds9.py]: set a fits file with ``set_fits`` and then retrieve it usind ``xpaget``. They should be the same

**Comment by ericmandel** [comment]
@montefra Better than good news, this is great news and great work. The whole Python community owes you a debt ...

The changes looks fine and make sense, although I hasten to admit that if there are subtle details, they may be a bit beyond me, my Python knowledge having leaked away due to non-use.

Again, it's really good work.

**Comment by montefra** [comment]
> The changes looks fine and make sense, although I hasten to admit that if there are subtle details, they may be a bit beyond me, my Python knowledge having leaked away due to non-use.

The main thing is to check that I am not doing something stupid with the way I communicate with ds9, in particular when using ``xpaget``.

> Again, it's really good work.

Thank you :blush:

**Comment by montefra** [comment]
thank you for checking. Sometimes English slips ...

On the ``{get,set}_pyfits``: I tried to keep its behaviour as similar as possible to the old one. If ``pyfits`` is not installed, warn the user to use ``{get,set}_fits``. Otherwise ``{get,set}_pyfits`` gets or returns a ``pyfits.HDUList`` instance.

**Comment by montefra** [comment]
If you don't have any request or comment, I might merge this PR

**Comment by ericmandel** [comment]
It looks fine, I'd merge it and take the rest of the day off!

**Comment by montefra** [comment]
@ericmandel : do you ha ve preferences on how to do the merging?

The merge button shows three options:

1. ``Merge pull request``: I guess that it does a merge or a fast-forward depending on the status
2. ``Squash and merge``: I don't like the idea of squashing
3. ``Rebase and merge``: first make sure that the changes are on top of the master and then do a fast-forward.

I think that for this case 1. and 3. are the same

**Comment by ericmandel** [comment]
I generally just merge. Both rebasing and squashing can hide the actual flow of development and, in general, I would prefer to have the commits match that flow.

**Comment by montefra** [comment]
@ericmandel : fine with me.

**Pull request merged**